From 0aaee76f5e3bdff177eea513bb6cd30874adb82a Mon Sep 17 00:00:00 2001
From: Steve Lhomme <robux4@ycbcr.xyz>
Date: Tue, 6 Jun 2023 09:44:30 +0200
Subject: [PATCH 3/4] Use CreateFile2 when building for Win8+

CreateFileW is not allowed in Universal Windows Platform builds, but
CreateFile2 is. We call always use the call on Win8+ platforms which
includes all the UWP platforms and provides common code.
---
 src/device.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/src/device.c b/src/device.c
index deaf7db..ae39b1a 100644
--- a/src/device.c
+++ b/src/device.c
@@ -488,6 +488,7 @@ static int win2k_open ( dvdcss_t dvdcss, const char *psz_device )
      * won't send back the right result).
      * (See Microsoft Q241374: Read and Write Access Required for SCSI
      * Pass Through Requests) */
+#if _WIN32_WINNT < 0x0602 /* _WIN32_WINNT_WIN8 */
     h_fd =
                 CreateFileW( psz_dvd, GENERIC_READ | GENERIC_WRITE,
                             FILE_SHARE_READ | FILE_SHARE_WRITE,
@@ -499,7 +500,23 @@ static int win2k_open ( dvdcss_t dvdcss, const char *psz_device )
                     CreateFileW( psz_dvd, GENERIC_READ, FILE_SHARE_READ,
                                 NULL, OPEN_EXISTING,
                                 FILE_FLAG_RANDOM_ACCESS, NULL );
+#else
+    CREATEFILE2_EXTENDED_PARAMETERS params;
+    ZeroMemory(&params, sizeof(params));
+    params.dwSize = sizeof(params);
+    params.dwFileFlags = FILE_FLAG_RANDOM_ACCESS;
+    h_fd =
+                CreateFile2( psz_dvd, GENERIC_READ | GENERIC_WRITE,
+                            FILE_SHARE_READ | FILE_SHARE_WRITE,
+                            OPEN_EXISTING,
+                            &params );
 
+    if( h_fd == INVALID_HANDLE_VALUE )
+        h_fd =
+                    CreateFile2( psz_dvd, GENERIC_READ, FILE_SHARE_READ,
+                                OPEN_EXISTING,
+                                &params );
+#endif
     if( h_fd == INVALID_HANDLE_VALUE )
     {
         print_error( dvdcss, "failed to open device %s", psz_device );
-- 
2.37.3.windows.1

